crates/sui-framework/sources/dynamic_field.move

```diff
@@ -252,4 +293,105 @@ spec has_child_object_with_ty {
     // TODO: specify actual function behavior
 }
 
+// === Dynamic fields sharded by key-value types ===
+
+#[verify_only]
+/// A slice of the dynamic fields associated with an object sharded by K type.
+struct NameShard<Name: copy + drop + store> has key {
```

Comment:
Do we need to model both of these, or can we rely on the keys in `PairShard` to supplant `NameShard`?

Reply:
Unfortunately, yes, we need to have both.

The reason is that not all keys in the same `NAME` type have the same `VALUE` type. We might have
- `k1: u64` mapped to `v1: bool` and
- `k2: u64` mapped to `b2: vector<u8>`.

In this case, `k1` and `k2` will be in the same `NameShard` but different `PairShard`.